watertap/unit_models/electroNP_ZO.py

```diff
@@ -0,0 +1,494 @@
+            raise ConfigurationError(
+                "The ElectroNP model was expecting a solvent and did not receive it."
+            )
+
+        if (
+            len(self.config.property_package.solute_set) == 0
+            and len(self.config.property_package.ion_set) == 0
+        ):
+            raise ConfigurationError(
+                "The ElectroNP model was expecting at least one solute or ion and did not receive any."
+            )
+
+    def build(self):
+        # Call UnitModel.build to setup dynamics
+        super().build()
+
+        self.scaling_factor = Suffix(direction=Suffix.EXPORT)
+
+        units_meta = self.config.property_package.get_metadata().get_derived_units
+
+        # Check configs for errors
+        self._process_config()
+
+        # Create state blocks for inlet and outlets
+        tmp_dict = dict(**self.config.property_package_args)
+        tmp_dict["has_phase_equilibrium"] = False
+        tmp_dict["defined_state"] = True
+
+        self.properties_in = self.config.property_package.build_state_block(
+            self.flowsheet().time, doc="Material properties at inlet", **tmp_dict
+        )
+
+        tmp_dict_2 = dict(**tmp_dict)
+        tmp_dict_2["defined_state"] = False
+
+        self.properties_treated = self.config.property_package.build_state_block(
+            self.flowsheet().time,
+            doc="Material properties of treated water",
+            **tmp_dict_2,
+        )
+        self.properties_byproduct = self.config.property_package.build_state_block(
+            self.flowsheet().time,
+            doc="Material properties of byproduct stream",
+            **tmp_dict_2,
+        )
+
+        # Create Ports
+        self.add_port("inlet", self.properties_in, doc="Inlet port")
+        self.add_port(
+            "treated", self.properties_treated, doc="Treated water outlet port"
+        )
+        self.add_port(
+            "byproduct", self.properties_byproduct, doc="Byproduct outlet port"
+        )
+
+        # Add isothermal constraints
+        @self.Constraint(
+            self.flowsheet().config.time,
+            doc="Isothermal assumption for treated flow",
+        )
+        def eq_treated_isothermal(b, t):
+            return b.properties_in[t].temperature == b.properties_treated[t].temperature
+
+        @self.Constraint(
+            self.flowsheet().config.time,
+            doc="Isothermal assumption for byproduct flow",
+        )
+        def eq_byproduct_isothermal(b, t):
+            return (
+                b.properties_in[t].temperature == b.properties_byproduct[t].temperature
+            )
+
+        # Add performance variables
+        self.recovery_frac_mass_H2O = Var(
+            self.flowsheet().time,
+            initialize=0.8,
+            domain=NonNegativeReals,
+            units=pyunits.dimensionless,
+            bounds=(0.0, 1.0000001),
+            doc="Mass recovery fraction of water in the treated stream",
+        )
```

Comment:
Now that we're making this its own unit, separate from the ZO model framework, I wonder if we even need this recovery variable here. If so, I'd think the value would be very close to 1.

Reply:
I choose to fix the recovery to 1.